Stop using ES search_type=scan

joxl opened the pull request:
Stop using `search_type=scan`

- deprecated since `2.1`
- see https://www.elastic.co/guide/en/elasticsearch/reference/2.1/breaking_21_search_changes.html

Notable changes:

- Implements `scroll()` queries with identical "python processing" to that of `search()` queries, with new tests to verify that the results for both types of queries must be identical.
  - Fixes "scroll _id bug" which caused `scroll()` results to lack the top-level document `_id` attribute.
- Adds new tests for `corehq.elastic` and `corehq.util.es`.
- Removes `corehq.elastic` dependency on `corehq.pillows` module.
- Removes `corehq.util.es.interface` dependency on `pillowtop` module.
- Implements new `ES_META` registry, allowing tests and other code to use adhoc index/alias names without hard-coding them in `interface.py`.
- Fixes bugs in `reindexer.py` and `test_xform_management.py`.

Jira: [SAAS-12530](https://dimagi-dev.atlassian.net/browse/SAAS-12530)

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

- New tests added for interface changes.
- Some existing flaky tests have been fixed.
- New "ES_META registry" implementation makes adding ES tests easier.

### QA Plan

QA requested: [QA-3551](https://dimagi-dev.atlassian.net/browse/QA-3551)

### Safety story

Using alternative (now recommended) mechanism for performing a "scan" request. Current version of ES (2.4) performs searches efficiently in this way. No functionality is removed.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

joxl committed:
Stop using ES `search_type=scan`

- deprecated since 2.1
- see: https://www.elastic.co/guide/en/elasticsearch/reference/2.1/breaking_21_search_changes.html

millerdev commented on corehq/util/es/interface.py:
```diff
@@ -113,7 +113,8 @@ def scroll(self, scroll_id=None, body=None, params=None, **kwargs):
     def scan(self, index_alias, query, doc_type):
-        return scan(self.es, query=query, index=index_alias, doc_type=doc_type, search_type='scan')
+        return scan(self.es, query=query, index=index_alias, doc_type=doc_type,
+                    preserve_order=True)
```
Does `preserve_order=True` replace `search_type='scan'`? The [change log](https://www.elastic.co/guide/en/elasticsearch/reference/2.1/breaking_21_search_changes.html) mentions sorting by `_doc`; is that equivalent?

joxl replied:
Hmm, I thought this was the case (using `preserve_order=True` resulted in a `sort='_doc'` query), but while fetching the requisite links to confirm my understanding, I realized a mistake.

The docs for this function are extremely misleading.  Even to this day (as of version 7.x) they [reference setting `search_type` to `scan`](https://elasticsearch-py.readthedocs.io/en/v7.14.0/helpers.html?highlight=search_type#scan), despite Elasticsearch not _having_ a `scan` search type since v2.0.  I little bit of digging turned up two, two-year-old GitHub issues regarding the matter (one [still open](https://github.com/elastic/elasticsearch-py/issues/842) and one [closed](https://github.com/elastic/elasticsearch-py/issues/835)).

I think the python library completely changed the functionality of this `preserve_order` argument from 5.x onward without changing the docs. When set to `False` (the default) in any library version <5.x, the implementation [adds the `search_type='scan'` query parameter](https://github.com/elastic/elasticsearch-py/blob/2.5.0/elasticsearch2/helpers/__init__.py#L275-L276) (even though that parameter was deprecated in ES v2.1).

In all library versions >=5.x, the implementation now [explicitly sets the query argument to `sort='_doc'`](https://github.com/elastic/elasticsearch-py/blob/5.5.5/elasticsearch5/helpers/__init__.py#L359-L361), silently clobbering any existing `query` argument specified by the caller.

I now think that python library never stopped using `search_type='scan'` in any 2.x version. It appears as though they only dropped it in v5.x, despite Elasticsearch providing an alternative way to achieve the same (performant) end result without this parameter since v2.1.  If there is no other way to achieve this without forking the `scan()` function into the HQ repo and re-writing it (not worth the effort), then I think this PR is pointless. I'll have a look.

joxl committed:
Implement use of elasticsearch.helpers.scan in a version-agnostic way.

millerdev commented on corehq/util/es/interface.py:
```diff
@@ -113,6 +115,26 @@ def scroll(self, scroll_id=None, body=None, params=None, **kwargs):
+        In library versions >=5.x this results in the query['sort'] getting
+        silently clobbered to '_doc'.
+
+        To handle this appropriately across versions >=2.x:
```
Would it be more correct to say `>=2.1`? (or whatever version >2.0 where `search_type='scan'` was introduced)

joxl committed:
Add tests for previous commit

inb4 this is backwards

esoergel commented on corehq/util/es/interface.py:
```diff
@@ -113,7 +115,28 @@ def scroll(self, scroll_id=None, body=None, params=None, **kwargs):
         return results
 
     def scan(self, index_alias, query, doc_type):
```
I think it might be better to just stop using scan entirely and switch to scroll now.  It looks like the only place this is used is in the tragically named `scroll_query`:
https://github.com/dimagi/commcare-hq/blob/e69bf0e66ff3621662f8bea85a9cec7ba3e3ba50/corehq/elastic.py#L235-L238
This in turn is only used in `ESQuery.scroll`:
https://github.com/dimagi/commcare-hq/blob/e69bf0e66ff3621662f8bea85a9cec7ba3e3ba50/corehq/apps/es/es_query.py#L235-L239

This definitely expects to support sorting.  In fact, I just checked and saw at least two places that call that with a sort param - notably exports.  Does this mean that that param has just been ignored thus far?  Could that `ESQuery.scroll` method instead be switched to use an actual scroll query?  If a sort param isn't provided, we should be able to default to `_doc`, which would preserve the behavior for unordered queries.

This would make the query changes happen now, and not as part of a library upgrade, though I think that's actually an advantage.

joxl replied:
Thanks @esoergel. Looking into that `scroll_query` usage was at the top of my todo list for today.  I've been suspicious of the fact that `ESQuery.scroll` (and thus, most/all ES models) is based on this `scan()` helper. Certainly needs further investigation.

> Does this mean that that param has just been ignored thus far?

This is a possibility, but I have not confirmed it.  I have not tested what happens in a _sorted_ search query where the `search_type='scan'` parameter is added (this is the current HQ implementation of `ESQuery.scroll`).  Using `search_type='scan'` seems to imply an unordered result, so I don't know what the ES engine does when you specify it on a query with `sort`.

Upon switching to the `elasticsearch5` library (without this PR), any use of (the current implementation of) `ESQuery.scroll` to perform a sorted query will resulted in the `sort` field being silently clobbered to `_doc` due to our `scroll_query`'s use of the `scan()` helper.

> Could that ESQuery.scroll method instead be switched to use an actual scroll query?

Yes, and it should be. I believe it has always been true that using `sort` in a query with the `scan()` function negates the benefits of that helper (because specifying `search_type='scan'` implies an unordered result).

joxl committed:
Add ES interface tests in preparation for removing scan()

- verify scroll_query() results (as-built)
- verify scan() results "_id bug"

joxl committed:
Move test_es_interface into dedicated ES tests directory

joxl committed:
Replace `scan()` with `scroll()`

- fixes scan() results "_id bug"

joxl committed:
Fix import bug in pillowtop `reindexer.py`

joxl committed:
Implement global ES_META registry

joxl committed:
Fix ES state bugs in `test_xform_management`

- prior to fix, test could never pass unless another test first created
  the `test_hqusers` index for it.
- test now cleans up ES indices after itself

joxl committed:
Merge branch 'master' into jm/remove-es-scan

joxl committed:
Add notest decorator for new testing util functions

joxl committed:
Add ES alias (de)registration for reports esaccessors tests

millerdev reviewed (commented):
:tada:

joxl commented on corehq/util/es/interface.py:
```diff
@@ -20,12 +20,8 @@ def put_mapping(self, doc_type, mapping, index):
-            raise ESError(
-                f"{index_or_alias} is an unknown alias, query target must be one of {all_es_aliases}")
+        from corehq.elastic import verify_registered_alias
+        verify_registered_alias(index_or_alias)
```
I went back and forth on this when I wrote it. In the end I decided not to add a cache because "if it was good enough before...".  Also, I wrote this before I fully understood what this entire mechanism was for (why we still need it).  I forged forward with the commits on this PR as to not block further progress on async communication (the current implementation is not worse than what it replaces), but there is a good chance we can rip this entire registry logic out and not perform any of these checks at all.

joxl commented on corehq/apps/hqcase/tests/test_dbaccessors.py:
```diff
@@ -180,6 +180,7 @@ class ESAccessorsTest(TestCase):
 
     def setUp(self):
         super(ESAccessorsTest, self).setUp()
+        register_alias(CASE_INDEX_INFO.index, CASE_INDEX_INFO)
```
The `register_alias()` (and friends) are more verbose because there are tests that require "non-standard registries", and I was trying to keep this "registry change set" as small as possible (i.e. attempting to avoid overhauling every ES test out there).  I think another iteration of refactoring could make more improvements, which will happen. Right now there are ongoing communications about whether or not we still need this registry at all.

joxl committed:
Split elastic registry into `corehq.apps.es.registry`

joxl committed:
Consolidate `corehq.util.es` testing utils into `corehq.apps.es`

joxl committed:
Preemptive registry rename to reduce noise in next commit

joxl committed:
Privatize index registry dict and implement API instead

joxl committed:
Prevent `test_stale_data_in_es.py` from littering test output

joxl committed:
Move Elastic index "canonical name" definitions into `pillows.mappings`

dannyroberts commented:
As a matter of procedure I'm just checking back in to see if this PR is blocking I can help with or moving along still behind the scenes

joxl committed:
Extend `es_test` to handle elastic index registry boilerplate

joxl committed:
Improve Elastic tests using non-standard indices with new `es_test` decorator features

joxl committed:
Post self-review lint

millerdev reviewed (commented):
Looking great! I think only one blocking comment (generator test class method). The others are mainly style and naming suggestions.

joxl committed:
Use common nose attribute decorator

- provides better clarity for which tests are testing the `es_test` decorator
  in `test_test_utils.py`

joxl committed:
Fix `es_test` edge-case bugs

- Prevent bug resulting in AttributeError when used on test classes lacking
  setUpClass or tearDownClass methods.
- Raise ValueError instead of AttributeError if decorating classes with
  existing (poorly named) not-classmethod functions that collide with what
  `setup_class=True` expects.

joxl committed:
Review feedback: improve naming

joxl committed:
Review feedback: test that class methods are decorated and called

joxl committed:
Revert invalid test generator

joxl committed:
Fixup decorator tests

joxl committed:
Review feedback: improve tests

joxl committed:
Merge branch 'master' into jm/remove-es-scan

proteusvacuum committed:
Merge remote-tracking branch 'origin/master' into jm/remove-es-scan
